pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,155 @@
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsclient "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var decoder runtime.Decoder
+
+func init() {
+	scheme := runtime.NewScheme()
+	_ = gardencorev1alpha1.AddToScheme(scheme)
+
+	decoder = serializer.NewCodecFactory(scheme).UniversalDecoder()
+}
+
+func createEnqueueOnAddFunc(queue workqueue.RateLimitingInterface) func(extensionObject interface{}) {
+	return func(newObj interface{}) {
+		extensionObject := newObj.(extensionsv1alpha1.Object)
+		if extensionObject.GetExtensionStatus().GetState() == nil {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func createEnqueueOnUpdateFunc(queue workqueue.RateLimitingInterface) func(newExtensionObject, oldExtensionObject interface{}) {
+	return func(newObj, oldObj interface{}) {
+		var (
+			newExtensionObj = newObj.(extensionsv1alpha1.Object)
+			oldExtensionObj = oldObj.(extensionsv1alpha1.Object)
+		)
+
+		if newExtensionObj.GetExtensionStatus().GetState() == nil {
+			return
+		}
+
+		if oldExtensionObj.GetExtensionStatus().GetState() == nil {
+			enqueue(queue, newObj)
+		}
+
+		if apiequality.Semantic.DeepEqual(newExtensionObj.GetExtensionStatus().GetState().Raw, oldExtensionObj.GetExtensionStatus().GetState().Raw) {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueue(queue workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	queue.Add(key)
+}
+
+func (s *SyncController) createShootStateSyncReconcileFunc(kind string, obj runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		err := s.seedClient.Client().Get(s.context, req.NamespacedName, obj)
+		if apierrors.IsNotFound(err) {
+			s.log.Debug("Skipping because resoruce with kind was deleted")
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		extensionObject, err := apiextensions.Accessor(obj)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		state := extensionObject.GetExtensionStatus().GetState()
+		if state == nil {
+			return reconcile.Result{}, fmt.Errorf("invalid state. extension state for resource with kind %s in namespace %s was nil", kind, req.Namespace)
+		}
+
+		shoot, err := fromCluster(req.Namespace, s.extensionsSeedClient)
```

Comment:
Instead of reading the `Cluster` resource, can't we have a reverse function of https://github.com/gardener/gardener/blob/e7cf5dd21032ce3b3bd8d6715db7bb17ab398a84/pkg/operation/shoot/shoot.go#L219-L231 and use this?

Reply:
Would definitely be simpler. Thanks for the suggestion.

Reply:
We somehow have to get the Shoot name and project name only by having the extension name and namespace. That's why I've decided to pull it out from the Cluster resource. As far as I understand you suggest that I extract the project name from the Shoot's technical id ?

Reply:
Since the shoot's namespace is always shoot--project-name--shoot-name it shouldn't be a problem to just get the shoot-name. You will have to define the function close to ComputeTechnicalID so that if it changes the reverse function should also change.

Reply:
Exactly as @plkokanov outlined :)